pkg/repository/prisma/dbsqlc/step_runs.sql

```diff
@@ -846,3 +863,51 @@ FROM
     "StepRunResultArchive"
 WHERE
     "stepRunId" = @stepRunId::uuid;
+
+
+-- name: ClearStepRunPayloadData :one
+WITH deleted_count AS (
+    SELECT COUNT(*) as count
+    FROM "StepRun" sr1
+    WHERE
+        sr1."tenantId" = @tenantId::uuid AND
+        sr1."deletedAt" > NOW() + INTERVAL '5 minutes'
+        AND (sr1."input" IS NOT NULL OR sr1."output" IS NOT NULL OR sr1."error" IS NOT NULL)
+), deleted_with_limit AS (
+    SELECT
+        sr2."id"
+    FROM "StepRun" sr2
+    WHERE
+        sr2."tenantId" = @tenantId::uuid AND
+        sr2."deletedAt" > NOW() + INTERVAL '5 minutes'
+        AND (sr2."input" IS NOT NULL OR sr2."output" IS NOT NULL OR sr2."error" IS NOT NULL)
+    ORDER BY "deletedAt" ASC
+    LIMIT sqlc.arg('limit')
+    FOR UPDATE SKIP LOCKED
+), deleted_archives AS (
+    SELECT sra1."id" as "id"
+    FROM "StepRunResultArchive" sra1
+    WHERE
+        sra1."stepRunId" IN (SELECT "id" FROM deleted_with_limit)
+        AND (sra1."input" IS NOT NULL OR sra1."output" IS NOT NULL OR sra1."error" IS NOT NULL)
+), cleared_archives AS (
+    UPDATE "StepRunResultArchive"
+    SET
+        "input" = NULL,
+        "output" = NULL,
+        "error" = NULL
+    WHERE
+        "id" IN (SELECT "id" FROM deleted_archives)
+)
+UPDATE
+    "StepRun"
+SET
+    "input" = NULL,
+    "output" = NULL,
+    "error" = NULL
+WHERE
+    "id" IN (SELECT "id" FROM deleted_with_limit)
+RETURNING
+    (SELECT count FROM deleted_count) as total,
+    (SELECT count FROM deleted_count) - (SELECT COUNT(*) FROM deleted_with_limit) as remaining,
```

Comment:
I'm concerned about the logic of returning remaining - I'm curious if there's a chance that the counts might not align if new data is pushed while executing this query. Are we guaranteed that both counts are seeing the same snapshot of the data?

Additionally, the count across all `deleted_count` is not ideal. Could we modify these queries so that we return `has_more` as a boolean, call `LIMIT sqlc.arg('limit') + 1`, and just case on whether there's at least 1 more step run than the limit? Since we're not using `remaining` for anything other than signifying whether to continue the deletion loop.

Reply:
good call, revised